File: src/adapters/mastercardAdapter.js
'use strict';

const { PaymentAuthorizationConfigurationError } = require('../errors/PaymentAuthorizationConfigurationError');

const ACCEPTED_PROFILE_IDS = ['card-default'];

/**
 * Mastercard adapter. Certified against the shared default profile.
 */
function authorize(payment, profile) {
  if (!profile || !ACCEPTED_PROFILE_IDS.includes(profile.id)) {
    throw new PaymentAuthorizationConfigurationError(
      'Mastercard requires a profile certified against the default acquirer BIN',
      {
        cardNetwork: 'mastercard',
        selectedProfile: profile ? profile.id : null,
        expectedProfile: ACCEPTED_PROFILE_IDS[0],
      },
    );
  }

  return {
    network: 'mastercard',
    profileId: profile.id,
    captureMode: profile.captureMode,
    installmentPlan: payment.installmentPlan || null,
    authorizationCode: `MC${String(payment.amountMinor).padStart(6, '0')}`,
    approved: true,
  };
}

module.exports = { authorize, ACCEPTED_PROFILE_IDS };

File: src/main/kotlin/com/beaconstone/payments/adapters/MastercardAdapter.kt
package com.beaconstone.payments.adapters

import com.beaconstone.payments.config.AuthProfile
import com.beaconstone.payments.errors.PaymentAuthorizationConfigurationError
import com.beaconstone.payments.services.Payment

private val ACCEPTED_PROFILE_IDS = setOf("card-default")

/** Mastercard adapter. Certified against the shared default profile. */
object MastercardAdapter : CardNetworkAdapter {
    override fun authorize(payment: Payment, profile: AuthProfile): AuthorizationResult {
        if (profile.id !in ACCEPTED_PROFILE_IDS) {
            throw PaymentAuthorizationConfigurationError(
                "Mastercard requires a profile certified against the default acquirer BIN",
                cardNetwork = "mastercard",
                selectedProfile = profile.id,
                expectedProfile = "card-default",
            )
        }

        return AuthorizationResult(
            network = "mastercard",
            profileId = profile.id,
            captureMode = profile.captureMode,
            authorizationCode = "MC" + payment.amountMinor.toString().padStart(6, '0'),
            approved = true,
            acquirerBin = profile.acquirerBin,
            installmentPlan = payment.installmentPlan,
        )
    }
}
